### DISCUSSION TITLE
Python FFI Support

### DISCUSSION THREAD

**Discussion by mplemay** [oven-sh/bun#10775] [Ideas]
Bun (and the typescript ecosystem are amazing); together, developer are able to quickly build out amazing full-stack web application. Yet, when creating non-trivial applications which require more than just some basic CRUD logic, the typescript full-stack ecosystem doesn't quite live up to hype. Increasingly, in order to build out AI driven applications (and more) developers have to turn to another language (like python) and create an API which wraps some bit of functionality.

Creating a REST API works, but as most know, it doesn't work well (even when using OpenAPI codegen tools). To get these applications working, developers have to attempt to sync state, duplicate database models, and undertake many other boilerplate tasks across their "frontend" and "backend" codebases. While this was a necessity for 2010s when developers moved from SSR rendering to SPA apps, we are seeing a renaissance of SSR (this time with hydration). Therefore, if we are back to rendering on the server, why do we need still setup a bunch of bloatware in order to create a real application. Why don't we just start calling the "backend" directly?

While there are a bunch of interesting ways to accomplish this, arguably the best way is through Foreign Function Interfaces (FFIs). While bun does have a FFI, most most backends in use today (namely python) do not easily compile down to the de-facto FFI format (i.e. C). And while there is a way to call python from bun (i.e. https://github.com/codehz/bun_python) or using wasm (i.e. pyodide), these methods do not have first class support within bun.

Given this, I was wondering if the bun team has any plans in the near or short term to support calling python code within bun?

**Comment by nexusprime-here** [comment]
Wtf, and how did you think we could do that?

**Reply by mplemay** [reply]
> And while there is a way to call python from bun (i.e. https://github.com/codehz/bun_python) or using wasm (i.e. pyodide), these methods do not have first class support within bun.

^ These two different projects demonstrate two different approaches to the problem. I assume the first approach I mentioned creates a python interpreter in the background using https://docs.python.org/3/c-api/index.html. And a pyodide based approach could rely on https://bun.sh/docs/runtime/loaders#wasm to run python.

**Comment by ghost** [comment]
> these methods do not have first class support within bun

I believe that they should not

**Comment by vjpr** [comment]
I think this is a great idea. Everytime I want to interop with Python for AI / data engineering, its always a pain.

Bun has a history of this with things like the sqlite integration and it fits with the philosophy of focusing on practical real-world dev experience above all.

**Comment by guest271314** [comment]
Check out what Cloudflare's Workerd is doing with Pyodide https://github.com/cloudflare/workerd/blob/35ec39b254dbde908e1d6428fb42c1615aa5bf98/src/pyodide/README.md.

**Reply by mplemay** [reply]
Thanks for sending this along! `pyodide` is definitely one option for running python code within js ecosystem. I would be curious to see if python could be integrated directly (without using `pyodide`) now that there is supposedly tier 2 wasm support (https://github.com/python/cpython/blob/main/Tools/wasm/README.md).

**Comment by Yuzu02** [comment]
I think it's an experiment, but someone is already testing it. 

https://github.com/oven-sh/bun/tree/dylan/python

Examples of the implementation:
https://x.com/dylanconway111/status/2003273218849931558

**Reply by leanderriefel** [reply]
are these just experimentations or is it actually planned to ship in the near future?
